[doc] Run A Flux Minicluster (#731)
kubernetes-sigs/kueue: site/content/en/docs/tasks/run_flux.md

Reviewer: run_flux_minicluster.md

---

> title: "Run A Flux Minicluster"

Reviewer 2: yes correct - `MiniCluster` !

> title: "Run A Flux Minicluster"

Reviewer 2: I'm going through our docs now to make sure all references are consistent - that will make it easier in the future to derive the right term.

> title: "Run A Flux Minicluster"

Author: done

date: 2022-02-14
weight: 6
description: >
  Run a Kueue scheduled Flux minicluster.
---

This page shows how to leverage Kueue's scheduling and resource management capabilities when running [Flux Operator's](https://flux-framework.org/flux-operator/) Mini Clusters.

> For more information on Flux Operator visit [Flux Operator](https://flux-framework.org/flux-operator/).

Reviewer: I think we can remove this line, given that there is already a link above.

But we should add:

The intended audience for this page are [batch users](https://kueue.sigs.k8s.io/docs/tasks#batch-user).

## Before you begin

> Please check [Run A Job](/docs/tasks/run_jobs) for details on the initial cluster setup and queue configuration.

Reviewer: Rather, point to `/docs/tasks/administer_cluster_quotas/`

Reviewer: oh, and another requirement is that the flux operator is installed :)

> Please check [Run A Job](/docs/tasks/run_jobs) for details on the initial cluster setup and queue configuration.

Reviewer 3: ```suggestion
Check [Run A Job](/docs/tasks/run_jobs) for details on the initial cluster setup and queue configuration.
```

## 1. MiniCluster definition

> Since a Flux MiniCluster uses  a `batch/job` to run the actual workload, no extra components are needed for it to be managed by Kueue, However the following aspects should be taken into account:

Reviewer: ```suggestion
Since a Flux MiniCluster runs as a [`batch/Job`](https://kubernetes.io/docs/concepts/workloads/controllers/job/), no extra components are needed for it to be managed by Kueue. However, the following aspects should be taken into account:
```

### a. Queue selection

> The target local queue should be specified in the `spec.jobLabels` section of the minicluster configuration.

Reviewer: link to /docs/concepts/local_queue

```yaml
  jobLabels:
    kueue.x-k8s.io/queue-name: user-queue
```

### b. Configure the resource needs

> The resource needs of the workload can be configured in the `spec.container[].resources` sections of the minicluster configuration.

Reviewer 2: If the user is running something with MPI or similar and they want more control, they also might want to consider tweaking `spec.container[].cores`.  or `spec.tasks` https://flux-framework.org/flux-operator/getting_started/custom-resource-definition.html?h=tasks#tasks. In layman's terms these are telling the exact resources that should be used for the single command. As an example from our kubecon experiment run, on a size 64 cluster:

```yaml
apiVersion: flux-framework.org/v1alpha1
kind: MiniCluster

metadata:
  name: lammps
  namespace: flux-operator
spec:
  # localDeploy needs to be false
  localDeploy: false

  # Number of pods to create for MiniCluster
  size: 64
  tasks: 6016

  # don't suppress output (in case there is issue) and time flux
  logging:
    quiet: false
    timed: true
    debug: true

  # This is a list because a pod can support multiple containers
  containers:
    - image: ghcr.io/rse-ops/spack-ubuntu-libfabric-ssh:ubuntu-20.04
      workingDir: /home/flux/examples/reaxff/HNS
      command: lmp -v x 64 -v y 16 -v z 16 -in in.reaxc.hns -nocite

      # Option Flags for this flux runner wait.sh entrypoint
      fluxOptionFlags: "-ompi=openmpi@5 -c 1 -o cpu-affinity=per-task" 

      # Leave 2 cores for kubernetes, we have 96,
      # this is just for eksctl
      cores: 96

      # Resource limits to enable efa
      resources:
        limits:
          vpc.amazonaws.com/efa: 1
          memory: "340G"
          cpu: 94

        requests:
          vpc.amazonaws.com/efa: 1
          memory: "340G"
          cpu: 94

      # custom preCommand logic (run at start of script)
      preCommand: |
        source /etc/profile.d/z10_spack_environment.sh
        asFlux="sudo -u flux -E PYTHONPATH=$PYTHONPATH -E PATH=$PATH -E FI_EFA_USE_DEVICE_RDMA=1 -E RDMAV_FORK_SAFE=1"
        . /etc/profile.d/z10_spack_environment.sh 
        cd /opt/spack-environment
        . /opt/spack-environment/spack/share/spack/setup-env.sh
        spack env activate .
        cd /home/flux/examples/reaxff/HNS
```
That generated the following command for flux start:

```bash
🌀 flux start $brokerOptions flux mini submit  -N 64 -n 6016 --quiet -ompi=openmpi@5 -c 1 -o cpu-affinity=per-task --watch -vvv lmp -v x 64 -v y 16 -v z 16 -in in.reaxc.hns -nocite
```
Note that `-n` is 6016 and `-N` is 64, and in flux this maps to:

```bash
  -N, --nodes=N               Number of nodes to allocate
  -n, --ntasks=N              Number of tasks to start
```

Reviewer: For the purpose of queuing, we care about how many resources the pods will consume in the cluster.
We will obtain those from the created Job object.

> The resource needs of the workload can be configured in the `spec.container[].resources` sections of the minicluster configuration.

Reviewer 4: ```suggestion
The resource needs of the workload can be configured in the `spec.container[*].resources` sections of the minicluster configuration.
```

It would be good to do the same as other places.
example: https://kueue.sigs.k8s.io/docs/tasks/administer_cluster_quotas/#2-create-a-clusterqueue-referencing-the-flavors

Reviewer 2: That might be hard to keep up to date / synced with our docs - so another suggestion is to just link to our documentation https://flux-framework.org/flux-operator/getting_started/custom-resource-definition.html#id3

Reviewer 4: My suggestion is to add `*` to the array :)

Reviewer 4: This means I didn't suggest referring to other documentation :)

Reviewer 2: Or sorry, it was hard to get the full context when you just said "do the same as other places" with a link. I did see the suggestion. And I think it still might be good to refer to the docs on our site!

```yaml
spec:
  containers:
    - image: <image>
      resources:
        requests:
          cpu: 4
          memory: "200Mi"
```

> ## 2. Configuration example

Reviewer: ```suggestion
## 2. Sample MiniCluster
```

```yaml
apiVersion: flux-framework.org/v1alpha1
kind: MiniCluster
metadata:
  generateName: flux-sample-kueue-
spec:
  size: 1
  containers:
    - image: ghcr.io/flux-framework/flux-restful-api:latest

>       command: echo hello world 

Reviewer: Is this a meaningful command @vsoch?

Is there a hello HPC? calculate pi, etc?

Reviewer 2: Oh yeah you mean:

```bash
echo hello hpc :wave: 
```
:laughing: 

We have quite a few examples https://github.com/flux-framework/flux-operator/tree/main/examples and actually the one I'd recommend (with a pre-built container that does something marginally interesting that uses MPI) is... LAMMPS!

https://github.com/flux-framework/flux-operator/blob/main/examples/tests/lammps/minicluster.yaml

We do have a hello-world example https://github.com/flux-framework/flux-operator/blob/main/examples/tests/hello-world/minicluster.yaml but I'd stay away from it, because it runs so quickly you'd probably miss it.

Author: switch to `sleep 10`, to simulate some work

      resources:
        requests:
          cpu: 4
          memory: "200Mi"
  jobLabels:
    kueue.x-k8s.io/queue-name: user-queue
```

## 3. Known Limitations

> - The certificate generation step of the MiniCluster life cycle runs outside it's main job and it's execution will not be managed by Kueue.

Reviewer: /hold
we might be able to remove this soon.

Reviewer 2: This has been merged so safe to remove. Anyone that installs the operator after this morning will not get a certificate pod!

> - The certificate generation step of the MiniCluster life cycle runs outside it's main job and it's execution will not be managed by Kueue.

Reviewer 3: ```suggestion
- The certificate generation step of the MiniCluster life cycle runs outside its main job and its execution will not be managed by Kueue.
```

Assuming these are referring to the attributes that belong to the certificate generation

> - In some configuration variants the MiniCluster's job might not run to completion blocking the resources of the associated queue until it's deletion.

Reviewer: ```suggestion
- In some configuration variants, the MiniCluster's job might not run to completion blocking the resources of the associated queue until it's deletion.
```

I don't think this will change anytime soon, or is it @vsoch?

Reviewer 2: I'm definitely up to try out different strategies - what we really need is to have the logic afforded by the new JobSet. My only idea was to go off of what you mentioned about parent/child relationships @alculquicondor  and make the broker pod a separate job that is launched first, and then have the other cluster workers somehow children? (is there an example of this somewhere I haven't done it before!) But that might be more effort / refactor than it's worth if we eventually want to go back to one IndexedJob with more structure.

Author: we could leave this in and remove it later on

Author: removed

> - In some configuration variants the MiniCluster's job might not run to completion blocking the resources of the associated queue until it's deletion.

Reviewer 3: its instead of it's

Author: done

> Both of them are under investigation and tracked by [150](https://github.com/flux-framework/flux-operator/issues/150) and [151](https://github.com/flux-framework/flux-operator/issues/151)

Reviewer 3: ```suggestion
Both of these known issues are under investigation and tracked by [150](https://github.com/flux-framework/flux-operator/issues/150) and [151](https://github.com/flux-framework/flux-operator/issues/151)
```

Author: removed
